src/recovery/mod.rs

```diff
@@ -563,8 +652,250 @@ impl RecoveryStore<u64, TdPyAny, TdPyAny> for SqliteRecoveryStore {
+    topic: String,
+}
+
+/// Keys in our Kafka-based key-value store.
+///
+/// Again use an enum to get easy serialization to / from bytes (which
+/// is all Kafka can actually read / write).
+#[derive(Debug, Serialize, Deserialize)]
+enum KafkaKey {
+    /// Store a state update for a stateful operator.
+    State {
+        step_id: StepId,
+        key: TdPyAny,
+        epoch: u64,
+    },
+    /// Save that the dataflow frontier has progressed.
+    DataflowFrontier,
+}
+
+impl KafkaKey {
+    fn to_bytes(&self) -> Vec<u8> {
+        // TODO: Figure out if there's a more robust-to-evolution way
+        // to serialize this key. If the serialization changes between
+        // versions, then recovery doesn't work. Or if we use an
+        // encoding that isn't deterministic.
+        bincode::serialize(self).expect("Error serializing state key")
+    }
+
+    fn from_bytes(bytes: &[u8]) -> Self {
+        bincode::deserialize(bytes).expect("Error deserializing state key")
+    }
+}
+
+/// Values our or Kafka-based key-value store.
+///
+/// Again use an enum to get easy serialization to / from bytes (which
+/// is all Kafka can actually read / write).
+#[derive(Debug, Serialize, Deserialize)]
+enum KafkaPayload {
+    /// The actual state update. If inside is `None` then the state
+    /// was deleted at this epoch.
+    State(Option<TdPyAny>),
+    /// A dataflow frontier value update.
+    DataflowFrontier(u64),
+}
+
+impl KafkaPayload {
+    fn to_bytes(&self) -> Vec<u8> {
+        bincode::serialize(self).expect("Error serializing state payload")
+    }
+
+    fn from_bytes(bytes: &[u8]) -> Self {
+        bincode::deserialize(bytes).expect("Error deserializing state payload")
+    }
+}
+
+impl KafkaRecoveryStore {
+    fn new(config: PyRef<KafkaRecoveryConfig>) -> Self {
+        let rt = SendWrapper::new(
+            tokio::runtime::Builder::new_current_thread()
+                .enable_all()
+                .build()
+                .unwrap(),
+        );
+
+        let hosts = config.hosts.clone();
+        let producer: FutureProducer = ClientConfig::new()
+            .set("bootstrap.servers", hosts.join(","))
+            .create()
+            .expect("Error building Kafka producer");
+
+        Self {
+            rt,
+            hosts,
+            producer,
+            topic: config.topic.clone(),
+        }
+    }
+}
+
+impl RecoveryStore<u64, TdPyAny, TdPyAny> for KafkaRecoveryStore {
+    fn load(&self) -> (Vec<(StepId, TdPyAny, u64, Option<TdPyAny>)>, u64) {
+        debug!(
+            "Loading recovery data from {:?} topic={}",
+            self.hosts, self.topic
+        );
+
+        let consumer: StreamConsumer = self.rt.block_on(async {
+            ClientConfig::new()
+                // TODO: Should we be generating a new consumer group
+                // per cluster execution? Or will the group offset
+                // reset?
+                .set("group.id", "bytewax-loader")
```

Comment:
I've been thinking about this more, and I actually think the correct answer at this point is to not use consumer groups at all in recovery loading. Consumer groups do 2 things for you: automatic balancing of partitions across consumers, and the ability for a new consumer to resume from where a previous consumer left off. That second feature we'd have to actively disable here: if a new Timely worker joined this dataflow midway through loading, then we'd already have to restart all loading because all workers would not have the correct state information, so we'd have to use a new consumer group or delete the old one. And since distributing N partitions over a known number of workers is pretty easy from worker index, we could re-create static load balancing without much work.

In a glorious future in which we have dynamic dataflow cluster resizing, then perhaps we could go back to consumer groups because we would want to retain the progress of the entire group, but we'd need to think about how those things interact and if we'd even want to allow resizing during loading and not just execution. Even in that case, the consumer group isn't exactly per-dataflow; it's like per-dataflow instance or attempt. If you are recovering after a second failure, you don't want to resume state loading from where the first failure happened.

Reply:
In any case, I think I'm going to punt on this in this PR and tackle this in a followup one that fixes multiple worker loading.